includes/class-signature.php

```diff
@@ -347,21 +346,57 @@ public static function get_remote_key( $key_id ) {
 	/**
 	 * Gets the signature algorithm from the signature header.
 	 *
-	 * @param array $signature_block The signature block.
+	 * @see https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures-12
+	 *
+	 * @param array    $signature_block The signature block.
+	 * @param resource $public_key      The public key resource.
 	 *
-	 * @return string|bool The signature algorithm or false if not found.
+	 * @return int|\WP_Error The signature algorithm or WP_Error if not found.
 	 */
-	public static function get_signature_algorithm( $signature_block ) {
+	public static function get_signature_algorithm( $signature_block, $public_key ) {
 		if ( ! empty( $signature_block['algorithm'] ) ) {
 			switch ( $signature_block['algorithm'] ) {
 				case 'hs2019':
+					$details = \openssl_pkey_get_details( $public_key );
+
+					switch ( $details['type'] ?? 0 ) {
+						case \OPENSSL_KEYTYPE_RSA:
+							$bits = $details['bits'] ?? 2048;
+
+							if ( $bits >= 4 * KB_IN_BYTES ) {
+								return \OPENSSL_ALGO_SHA512;
+							} elseif ( $bits >= 3 * KB_IN_BYTES ) {
+								return \OPENSSL_ALGO_SHA384;
+							} else {
+								return \OPENSSL_ALGO_SHA256;
+							}
+
+						case \OPENSSL_KEYTYPE_EC:
+							$curve_name = $details['ec']['curve_name'] ?? '';
+
+							// 3 levels switch statements are fine, right?
+							switch ( $curve_name ) {
+								case 'prime256v1':
+								case 'secp256r1':
+									return \OPENSSL_ALGO_SHA256;
+								case 'secp384r1':
+									return \OPENSSL_ALGO_SHA384;
+								case 'secp521r1':
+									return \OPENSSL_ALGO_SHA512;
+							}
+					}
+
+					return new \WP_Error( 'unsupported_key_type', 'Unsupported key type (only RSA and EC keys are supported).', array( 'status' => 401 ) );
+
```

Comment:
@vpzomtrrfrt I'm curious if you'd have any advice on how this could be improved to [switch `algorithm="hs2019"` support for wordpress-activitypub to "Yes"](https://arewehs2019yet.vpzom.click/)?

Reply:
I haven't followed this space as closely lately, the last I heard was that there was no consensus on how keys should indicate their supported algorithms